### PR TITLE
Make session serializer extensible

### LINKED ISSUE #1438: Make TaggedJSONSerializer extensible
My suggestion would be to add a method to the class that lets you register custom types, e.g. like this:

```
session_json_serializer.register_type('md', MultiDict)
```

This would use the `' md'` prefix when encountering a `MultiDict` object (`type(x) is MultiDict` as subclasses probably shouldn't be handled). To serialize/deserialize I'd simply use `__getstate__` and `__setstate__`.

### PR DESCRIPTION
**Pull request by davidism**
Continues #1452.

Refactors `flask.session.TaggedJSONSerializer` to a new `flask.json.tag` module. Creates a `JSONTag` interface and a `TaggedJSONSerializer.register` method to register new tags.

~~~python
from flask.json.tag import JSONTag

class TagOrderedDict(JSONTag):
    __slots__ = ('serializer',)
    key = ' od'

    def check(self, value):
        return isinstance(value, OrderedDict)

    def to_json(self, value):
        return [[k, self.serializer.tag(v)] for k, v in iteritems(value)]

    def to_python(self, value):
        return OrderedDict(value)

app.session_interface.serializer.register(TagOrderedDict, 0)
~~~

Since `register` takes an index, it can be used to override existing tags (although it would be slightly more efficient to subclass `TaggedJSONSerializer` instead). This can address #1908 by adding a higher-priority `dict` serializer that uses a different format to support non-string keys.

Closes #1438, #1511, #1908.

~Still needs:~

- [x] refactor to move the `serializer` argument to `self.serializer` instead. It was an attempt at optimization but didn't change anything.
- [x] refactor `TagDict` to split the logic between addressing #1511 and regular dict passthrough, as #1452 originally did. This was another optimization attempt that didn't help, it just made the code more confusing.
- [x] more tests
- [x] documentation

### CODE REVIEW

File: flask/json/tag.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+from flask._compat import iteritems, text_type
+from flask.json import dumps, loads
+
+
+class JSONTag(object):
+    """Base class for defining type tags for :class:`TaggedJSONSerializer`."""
+
+    __slots__ = ('serializer',)
+    key = None
+    """The tag to mark the serialized object with. If ``None``, this tag is
```
Comment (ThiefMaster) [inline flask/json/tag.py]: AFAIK the most common sphinx way of documenting a class-level variable is this:

```python
#: lorem ipsum
#: blah blah blah
foo = None
```

At least I find this "dangling" docstring quite ugly and I think some editors even highlight it as a no-op statement  and/or a docstring that's not where it should be/

File: flask/json/tag.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    only used as an intermediate step during tagging."""
+
+    def __init__(self, serializer):
+        """Create a tagger for the given serializer."""
+
```
Comment (ThiefMaster) [inline flask/json/tag.py]: only class-level docstrings are followed by an empty line

File: flask/json/tag.py (hunk 1 above)
Comment (davidism) [inline flask/json/tag.py]: Sphinx and PyCharm both recognize it. Not a big deal though, I can switch to `#:` and ignore PyCharm.

**Comment by davidism** [comment]
@ThiefMaster from https://github.com/pallets/flask/pull/1452#issuecomment-305609953:

Re: performance: new method is ~ 30 µs slower.

~~~
old: 70.7 µs ± 164 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
new: 96.3 µs ± 678 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
~~~

You'd need to handle 2,000,000 requests before you wasted 1 extra minute. There's always Flask-Session, etc. where server-side sessions are serialized with `pickle`, which is *much* faster.

File: flask/json/tag.py (hunk 1 above)
Comment (ThiefMaster) [inline flask/json/tag.py]: ah ok, never noticed pycharm handling it nicely. probably never opened a file using this notation ;)

I just checked and `#:` is quite widespread in the flask codebase so I'd go for that

**Comment by ThiefMaster** [comment]
ok, that seems fine

**Pull request merged**